File: 07_다이나믹프로그래밍/07_병사배치하기.cpp
#include <iostream>
#include <vector>

using namespace std;

void showDP(vector<int>& _dp){
    cout << endl;
    for(auto i : _dp){
        cout << i << " ";
    }
    cout << endl;
}

int main(void){

    int n;
    cin >> n;

    vector<int> arr;
    for (int i = 0; i < n; i++)
    {
        int temp;
        cin >> temp;
        arr.push_back(temp);
    }
    // DP Table 1로 초기화
    vector<int> dp(n, 1); 
    int result = 0;

    for (int i = 1; i < n; i++)
    {
        for (int j = 0; j < i; j++)
        {
            // 내림차순을 만족할 때만 Dp Table update
            if(dp[j] > dp[i]){
                dp[i] = max(dp[i], dp[j] + 1);
                if(result > dp[i]) {
                    result = dp[i];
                }
            }
        }
    }

    showDP(dp);
    
    cout << result << endl;

    return 0;
}

/*

7
15 11 4 8 5 2 4

*/

File: 07_다이나믹프로그래밍/07_병사배치하기.py
def showDP(dp_table) :
    for i in dp_table :
        print(i, end=' ')
    print('n')


n = int(input())
arr = list(map(int, input().split()))

dp_table = [1] * n

# LDS 알고리즘 수행
for i in range(1, n) :
    for j in range(0, i) :
        # 내림차순이 성립하면, DP Table update
        if arr[j] > arr[i] :
            dp_table[i] = max(dp_table[i], dp_table[j] + 1)

showDP(dp_table)

# 열외해야 하는 병사의 최소 수   
print(n - max(dp_table))

'''

7
15 11 4 8 5 2 4

'''